### PR TITLE
Fix improve the review ahead refactor

### LINKED ISSUE #20600: Improve the Review ahead dialog
## The Review ahead dialog

Image

- [ ] Add a title "Review ahead".
- [ ] Replace the description with "Review cards due in the next:".
- [ ] Add a unit "day"/"days" to the box.
- [ ] Replace "OK" with "Create".
- [ ] If the resulting filtered deck is empty (i.e., no cards match the criteria), a red warning should appear and the button should be disabled, instead of an error dialog (see reference).
- [ ] Show a warning when the input is invalid, limit the digits, and prevent leading zeros, like in https://github.com/ankidroid/Anki-Android/pull/20520, except without input steppers.

## Reference dialogs

Image

Image

### PR DESCRIPTION
**Pull request by Galal-20**
## Purpose / Description
This PR improves the Review Ahead dialog UX to align with expected behavior and consistency
Additionally, the query construction has been refactored and clarified based on Anki's search syntax (deck + prop:due), which is used across Anki Desktop and backend.

## Fixes
Fixes: #20600

## Approach
The following improvements were implemented:

UI Enhancements:

	Added dialog title: “Review ahead”
	Updated description to: “Review cards due in the next:”
	Replaced positive button text from “OK” → “Create”
	Added dynamic unit suffix “day”/“days” 

Input Validation Improvements:

	Prevented leading zeros 
	Limited input length 
	Added real-time validation
	Disabled the action button when:
	input is invalid
	input is zero
	no matching cards are found

ْUnit Tests:

    Added unit tests to validate query construction logic:
    
    - Ensured correct query format for "Review Ahead":
        deck:"<deckName>" prop:due<=<daysAhead>
    
    - Ensured fallback behavior for other options:
        returns only deck:"<deckName>"
    
    - Tests cover edge cases and confirm consistency with Anki search syntax

## How Has This Been Tested?

Tested manually on:
Android Emulator (API 33) and Physical device

Test Cases:

	1.	Open “Review Ahead” dialog → verify:
	              Title is displayed
	              Description is correct
	              Default value shows correct suffix (“days”)

	4.	Input validation:
	              Enter non-numeric → error shown, button disabled
	              Enter 0 → error shown, button disabled
	              Enter valid number → suffix updates correctly (day/days)

	5.	Edge cases:
	              Enter large number → input limited correctly
	              Leading zero is prevented

	6.	Empty result case:
	             Enter value with no matching cards → inline error shown
	             Button disabled (no dialog appears)

## Screenshots
some screens: 

Reference:
https://docs.ankiweb.net/searching.html

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title.
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner]

### CODE REVIEW

**Comment by Galal-20** [comment]
Hello @david-allison,
Thanks for taking a look at this PR
I’ve refactored the “Review Ahead” dialog UX and query construction for clarity and consistency. 
All edge cases are covered with unit tests.
Happy to adjust anything based on your feedback.

**Comment by Haz3-jolt** [comment]
Hi, can you squash down your commits? We try to target for atomic commits.

Generally, try to aim for "type(scope): title" it helps keep the review process easy.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 1]
```diff
@@ -423,13 +433,117 @@ class CustomStudyDialog : AnalyticsDialogFragment() {
...
+    ) {
+        if (contextMenuOption != STUDY_AHEAD) return
+
+        if (userInputValue == null) {
+            binding.detailsEditText2Layout.error = null
+            dialog.positiveButton.isEnabled = false
+            return
+        }
+
+        if (userInputValue == 0) {
```
Comment (Haz3-jolt) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt]: This should be redundant with `class NoLeadingZeroFilter`

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 2]
```diff
@@ -423,13 +433,117 @@ class CustomStudyDialog : AnalyticsDialogFragment() {
...
+                R.plurals.set_due_date_label_suffix,
+                safeValue,
+                safeValue,
+            )
+
+        val currentInput = userInputValue
+        lifecycleScope.launch {
+            val deckName = withCol { decks.name(viewModel.deckId) }
+            val query = buildQuery(deckName, contextMenuOption, currentInput)
+            val matchingNotes = withCol { findNotes(query) }
```
Comment (Haz3-jolt) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:466]: I might be wrong, but wouldn't findNotes mean we end up effecting all cards associated?

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 3]
```diff
@@ -423,13 +433,117 @@ class CustomStudyDialog : AnalyticsDialogFragment() {
 
         binding.detailsEditText2.doAfterTextChanged {
             dialog.positiveButton.isEnabled = userInputValue != null && userInputValue != 0
+            val value = it?.toString()?.toIntOrNull()
+
+            // Prevent invalid inputs like leading zeros (e.g. "01") by normalizing the value.
+            if (replaceZeroWithNextNumber(it)) return@doAfterTextChanged
+            studyAheadCase(contextMenuOption, dialog, value)
```
Comment (Haz3-jolt) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt]: You should add a debounce here, or else every keystroke would end up triggering a separate coroutine.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 4]
```diff
@@ -423,13 +433,117 @@ class CustomStudyDialog : AnalyticsDialogFragment() {
...
+        val initialValue = defaultValue.toIntOrNull() ?: 1
+        binding.detailsEditText2Layout.suffixText =
+            resources.getQuantityString(
+                R.plurals.set_due_date_label_suffix,
+                initialValue,
+                initialValue,
+            )
+        binding.detailsText1.setTextSize(TypedValue.COMPLEX_UNIT_SP, 18f)
+    }
+    // Uses Anki search syntax to filter cards due within N days.
```
Comment (Haz3-jolt) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:510]: nit: Add a blank space here

File: AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt [hunk 5]
```diff
@@ -314,6 +315,64 @@ class CustomStudyDialogTest : RobolectricTest() {
...
 
+    @Test
+    fun `buildQuery returns correct string for STUDY_REVIEW_AHEAD`() =
+        runTest {
+            val deckName = "Default"
+            val daysAhead = 5
+
+            withCustomStudyFragment(args = argumentsDisplayingMainScreen()) { dialogFragment ->
+                val method =
+                    dialogFragment.javaClass.getDeclaredMethod(
```
Comment (Haz3-jolt) [inline AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt]: You can use @VisibleForTesting rather than reflections.

**Review by Haz3-jolt — CHANGES_REQUESTED** [review]
If possible try to add tests to cover the following cases
- Leading zero prevention                                                                                                                                                                                            - Zero input showing error                                                                                                                                                                                           - Button enabled/disabled states                                                                                                                                                                                     - Suffix updating (day/days)                                                                                                                                                                                         - Empty result disabling the button

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt (hunk 2 above)
Comment (Galal-20) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:466]: **findNotes** is used here strictly for UI validation (checking isNotEmpty() to toggle the Create button). 
No data is being modified at this stage, the actual card selection is handled later by the backend via **sched.customStudy**.
I used findNotes for a minor performance gain **fewer unique IDs**, but I'm happy to switch to findCards if you prefer consistency over this micro optimization. 
What do you think?

Comment (Haz3-jolt) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:466]: Then it is not a issue

Comment (Galal-20) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:466]: Glad we are on the same page

**Review by david-allison — CHANGES_REQUESTED** [review]
Thanks!

* Could you include the changes in the commit message
* Could you split this into two commits: one which handles the textual changes, and one which handles the input validation
  * this is harder to review as this concern is mixed in with the trivial refactors

**Comment by Galal-20** [comment]
Hello I'm sorry for replying late, can you open this PR again please?

File: AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt [hunk 6]
```diff
@@ -305,23 +305,102 @@ class CustomStudyDialogTest : RobolectricTest() {
     }
 
     @Test
-    fun `selectedKind maps selected card state index to cram kind`() {
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt:308]: These tests are deleted

Comment (Galal-20) [reply AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt:308]: Thanks for catching this, you are completely right.
I'm adding them back right now and will update the PR shortly.

**Comment by david-allison** [comment]
This has gone stale multiple times. CI is broken, and I still feel my last review needs to be handled.

I'll fix it up & credit you as I'd like to see it merged. Thanks for the effort!

**Comment by Galal-20** [comment]
> This has gone stale multiple times. CI is broken, and I still feel my last review needs to be handled.
> 
> 
> 
> I'll fix it up & credit you as I'd like to see it merged. Thanks for the effort!
> 
> 
> 
> 

Dear @david-allison 
Thank you for taking this over.
Everything was building fine locally on my end, so I wasn’t entirely sure what was causing the CI failure.
I really appreciate you taking the time to fix the CI, address the remaining review points, and make sure the work is credited appropriately.
Looking forward to seeing the changes merged.
Thanks again.
